### PR TITLE
Platops 1225 fix long lines

### PR DESCRIPTION
**Pull request by konradwudkowski**
- truncate lines if they exceed configurable limit
- no longer store value of a match as pointers to start and end are enough

### CODE REVIEW

File: test/uk/gov/hmrc/leakdetection/scanner/MatchedResultSpec.scala [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        description = "description",
+        matches     = matches)
+
+  def genMatch(lineText: String): Gen[Match] =
+    (for {
+      start <- Gen.choose(0, lineText.length)
+      end   <- Gen.choose(start, lineText.length)
+    } yield {
+      Match(start, end)
+    }).retryUntil(m => m.start != m.end)
```
Comment (hughfdjackson) [inline test/uk/gov/hmrc/leakdetection/scanner/MatchedResultSpec.scala]: Alternatively: 

```scala
  def genMatch(lineText: String): Gen[Match] =
    for {
      start <- Gen.choose(0, lineText.length)
      end   <- Gen.choose(start, lineText.length)
      if start != end
    } yield Match(start, end)
```

I understand that you get 100k retries with retryUntil, and you only get as many as scalatest's maxDiscards setting, so maybe this isn't exactly what you're after.

Comment (konradwudkowski) [reply test/uk/gov/hmrc/leakdetection/scanner/MatchedResultSpec.scala]: Yes, I more likely to not fail with ~100k~ 10k retries that `retryUntil` provides by default

File: app/uk/gov/hmrc/leakdetection/scanner/MatchedResult.scala [hunk 2]
```diff
@@ -25,25 +25,87 @@ final case class MatchedResult(
...
+      end   = regexMatch.end
     )
   implicit val format: Format[Match] = Json.format[Match]
 }
 
 object MatchedResult {
   implicit val format: Format[MatchedResult] = Json.format[MatchedResult]
+
+  def truncate(matchedResult: MatchedResult, limit: Int): MatchedResult =
+    if (matchedResult.lineText.length > limit && matchedResult.matches.nonEmpty) {
```
[the remark this answers is not in the document]
Comment (konradwudkowski) [reply app/uk/gov/hmrc/leakdetection/scanner/MatchedResult.scala]: good idea!

File: app/uk/gov/hmrc/leakdetection/scanner/MatchedResult.scala [hunk 3]
```diff
@@ -25,25 +25,87 @@ final case class MatchedResult(
...
+              (total, matches)
+            }
+        }
+
+      val joinedConsecutiveMatches =
+        matchesUpToLimit
+          .foldLeft(List.empty[Match]) {
+            case (lastAddedElement :: others, m) =>
+              if (lastAddedElement.end == m.start) {
+                lastAddedElement.copy(end = lastAddedElement.end + (m.end - m.start)) :: others
```
[the remark this answers is not in the document]
Comment (konradwudkowski) [reply app/uk/gov/hmrc/leakdetection/scanner/MatchedResult.scala]: yes, it's much better, thanks!

File: test/uk/gov/hmrc/leakdetection/scanner/MatchedResultSpec.scala (hunk 1 above)
Comment (hughfdjackson) [reply test/uk/gov/hmrc/leakdetection/scanner/MatchedResultSpec.scala]: I forgot the golden 'check before you comment' rule.  10k it is :)

File: app/uk/gov/hmrc/leakdetection/scanner/MatchedResult.scala [hunk 4]
```diff
@@ -25,25 +25,87 @@ final case class MatchedResult(
...
+
+      val (_, matchesUpToLimit) =
+        matchedResult.matches.foldLeft((0, List.empty[Match])) {
+          case ((total, matches), m @ Match(start, end)) =>
+            if (total + end - start <= limit) {
+              (total + end - start, matches :+ m)
+            } else {
+              (total, matches)
+            }
+        }
```
[the remark this answers is not in the document]
Comment (konradwudkowski) [reply app/uk/gov/hmrc/leakdetection/scanner/MatchedResult.scala]: :+1:

File: app/uk/gov/hmrc/leakdetection/scanner/MatchedResult.scala [hunk 5]
```diff
@@ -25,25 +25,87 @@ final case class MatchedResult(
...
+      val values = joinedConsecutiveMatches.map { m =>
+        matchedResult.lineText.substring(m.start, m.end)
+      }
+
+      val (_, matchesWithReadjustedIndexes) =
+        joinedConsecutiveMatches.zip(values).zipWithIndex.foldLeft(0, List.empty[Match]) {
+          case ((totalLength, acc), ((_, value), index)) =>
+            if (index == 0) {
+              val startPos = totalLength + "[…] ".length
+              val endPos   = totalLength + "[…] ".length + value.length
```
[the remark this answers is not in the document]
Comment (konradwudkowski) [reply app/uk/gov/hmrc/leakdetection/scanner/MatchedResult.scala]: `match.length` is the same as `value.length`, I will use it, it makes it simpler

File: test/uk/gov/hmrc/leakdetection/scanner/MatchedResultSpec.scala [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+          val initialTotalLengthOfAllMatches =
+            initialResult.matches.map(m => initialResult.lineText.substring(m.start, m.end).length).sum
+          if (initialTotalLengthOfAllMatches <= limit) {
+            val truncated = truncate(initialResult, limit)
+            initialResult.matches.foreach { m =>
+              val leakValue = initialResult.lineText.substring(m.start, m.end)
+              assert(truncated.lineText.contains(leakValue))
+            }
+          }
+      }
```
[the remark this answers is not in the document]
Comment (konradwudkowski) [reply test/uk/gov/hmrc/leakdetection/scanner/MatchedResultSpec.scala]: :)

File: app/uk/gov/hmrc/leakdetection/scanner/MatchedResult.scala [hunk 7]
```diff
@@ -25,25 +25,87 @@ final case class MatchedResult(
...
+  matches: List[Match],
+  isTruncated: Boolean = false
 )
 
 final case class Match(
   start: Int,
-  end: Int,
-  value: String
+  end: Int
 )
```
[the remark this answers is not in the document]
Comment (konradwudkowski) [reply app/uk/gov/hmrc/leakdetection/scanner/MatchedResult.scala]: good idea, added it

**Pull request merged**